sqldelight-runtime/src/main/java/com/squareup/sqldelight/Transacter.kt

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright (C) 2018 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.sqldelight
+
+import com.squareup.sqldelight.Transacter.Transaction
+import com.squareup.sqldelight.db.SqlDatabase
+
+/**
+ * A transaction-aware [SqlDatabase] wrapper which can begin a [Transaction] on the current thread.
+ */
+abstract class Transacter(
+  private val helper: SqlDatabase,
+  private val transactions: ThreadLocal<Transaction> = ThreadLocal()
+) {
+  /**
+   * For internal use, performs [function] immediately if there is no active [Transaction] on this
+   * thread, otherwise defers [function] to happen on transaction commit.
+   */
+  internal fun deferAction(function: () -> Unit) {
+    val transaction = transactions.get()
+    if (transaction != null) {
+      transaction.postCommitHooks.add(function)
+    } else {
+      function()
+    }
+  }
+
+  /**
+   * Starts a transaction and runs [body] in that transaction.
+   *
+   * @throws IllegalStateException if a [Transaction] is currently active on this thread.
+   * @see transaction
+   */
+  fun newTransaction(body: Transaction.() -> Unit) {
```

Comment:
`inTransaction`? `withTransaction`?

I don't like "new" when it doesn't return anything to you.

Reply:
I'm also not clear on when I would use this or `transaction`

Reply:
@swankjesse brought up the point that you may want to be able to know for certain that your changes were committed once the transaction block ends.

Reply:
I see. So `transaction` doesn't always nest. Is that an optimization?

Reply:
i think its necessary since sometimes you're unsure if you're in a transaction or not but want to run in a transaction. The way cash server guarantees this is by passing the transaction as one of the method parameters, which works but I like having the flexibility.

Reply:
I don't follow this at all.

The only thing this method does differently than calling `transaction` is throw when you're already in a transaction. What is the use case for this?

Reply:
```
fun foo() {
  transacter.transaction {
    // do some work.
  }
  // make a server call
}

...

fun bar() {
  transacter.transaction {
    foo()
    rollback()
  }
}
```

the work in `foo` doesnt get performed but the server call does, there might be cases where you dont want that to happen. Although writing that out I'm realizing you could instead use `afterCommit` inside the transaction inside foo and that would accomplish the same thing. Anyways thats the case.

Reply:
Ok. I'd be tempted to add some kind of boolean to accomplish that. The naming doesn't convey the difference to me. Something like
```
fun foo() {
  transacter.transaction(noParent = true) {
    // do some work.
  }
  // make a server call
}
```

Reply:
I wanted to call it `allowNesting` but that's confusing as to whether it can be nested or whether it can _contain_ nested...

Reply:
works for me 👍

Reply:
what about `isVisible`

but seriously I'm thinking `noEnclosing` to keep it consistent with the rest of the code

Reply:
Works for me. I was about to suggest `onlyRoot` or `mustBeRoot` as well.